test/k4FWCoreTest/src/components/ExampleFunctionalMetadataProducer.cpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2014-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Gaudi/Property.h"
+
+#include "edm4hep/MCParticleCollection.h"
+
+#include "k4FWCore/MetadataUtils.h"
+#include "k4FWCore/Producer.h"
+
+#include <string>
+#include <vector>
+
+struct ExampleFunctionalMetadataProducer final : k4FWCore::Producer<edm4hep::MCParticleCollection()> {
+  ExampleFunctionalMetadataProducer(const std::string& name, ISvcLocator* svcLoc)
+      : Producer(name, svcLoc, {}, KeyValues("OutputCollection", {"MCParticles"})) {}
+
+  StatusCode initialize() override {
+    k4FWCore::putParameter("NumberOfParticles", m_particleNum.value(), this);
+    k4FWCore::putParameter("ParticleTime", m_particleTime.value(), this);
+    k4FWCore::putParameter("PDGValues", m_PDGValues.value(), this);
+    k4FWCore::putParameter("MetadataString", m_metadataString.value(), this);
+    return StatusCode::SUCCESS;
+  }
+
+  edm4hep::MCParticleCollection operator()() const override {
+    k4FWCore::putParameter("EventMetadataInt", 5, this);
```

Comment:
Where do these parameters go? To the event frame or to the metadata frame? If the latter, I think this should be impossible. At least with the `MetaDataHandle` we went to some length to make this fail (at runtime), because if something is "true metadata" then it should be constant at file level. Hence, there should be no need to make it possible to even set this during event processing. If we can make this call go to the event frame, then it would be OK, I suppose, but it might be a bit confusing to have only one interface for doing both.

The notable exception for setting metadata during event processing is obviously the MarlinWrapper, but I don't think we should make this easily possible here, just because it's needed there.

Reply:
This is now forbidden the same way it is done for the `MetaDataHandle` so it should work with the MarlinWrapper.

Reply:
Now changed to prevent from putting parameters in the metadata frame as it was being done for the handle-based algorithms, it's literally the same mechanism so I think it should work for `MarlinWrapper`.